lib/inputstreamhelper/utils.py

```diff
@@ -73,143 +72,168 @@ def download_path(url):
+        checksum = None
 
-    req = _http_request(url)
-    if req is None:
+    response = _http_request(url)
+    if response is None:
         return None
 
     dl_path = download_path(url)
     filename = os.path.basename(dl_path)
-    if not message:  # display "downloading [filename]"
+    if not message:
         message = localize(30015, filename=filename)  # Downloading file
 
-    total_length = int(req.info().get('content-length'))
+    total_length = int(response.info().get('content-length', 0))
     if dl_size and dl_size != total_length:
         log(2, 'The given file size does not match the request!')
-        dl_size = total_length  # Otherwise size check at end would fail even if dl succeeded
+        dl_size = total_length
 
-    if background:
-        progress = bg_progress_dialog()
-    else:
-        progress = progress_dialog()
+    progress = _create_progress_dialog(background, message)
+
+    success = _download_file(response, dl_path, calc_checksum, total_length, message, progress, background)
+
+    progress.close()
+    response.close()
+
+    if not success:
+        return False
+
+    checksum_ok = _verify_checksum(checksum, calc_checksum)
+    size_ok = _verify_size(dl_size, dl_path)
+
+    if not checksum_ok or not size_ok:
+        if not _handle_corrupt_file(dl_size, dl_path, checksum, calc_checksum, filename):
+            return False
+
+    return dl_path
+
+
+def _initialize_checksum(checksum, hash_alg):
+    if not checksum:
+        return None
+
+    from hashlib import md5, sha1
+    if hash_alg == 'sha1':
+        return sha1()
+    if hash_alg == 'md5':
+        return md5()
+    log(4, 'Invalid hash algorithm specified: {}'.format(hash_alg))
+    return None
+
+
+def _create_progress_dialog(background, message):
+    progress = bg_progress_dialog() if background else progress_dialog()
     progress.create(localize(30014), message=message)  # Download in progress
+    return progress
 
+
+# pylint: disable=too-many-positional-arguments
+def _download_file(response, dl_path, calc_checksum, total_length, message, progress, background):
     starttime = time()
     chunk_size = 32 * 1024
+    size = 0
+
     with open(compat_path(dl_path), 'wb') as image:
-        size = 0
-        while size < total_length:
-            try:
-                chunk = req.read(chunk_size)
-            except (timeout, SSLError):
-                req.close()
-                if not yesno_dialog(localize(30004), '{line1}\n{line2}'.format(line1=localize(30064),
-                                                                               line2=localize(30065))):  # Could not finish dl. Try again?
-                    progress.close()
-                    return False
-
-                headers = {'Range': 'bytes={}-{}'.format(size, total_length)}
-                req = _http_request(url, headers=headers)
-                if req is None:
-                    return None
-                continue
+        while True:
```

Comment:
I try to avoid `while True`. To make sure we can never get stuck in an infinite loop, I would stick with `while size < total_length`